From d8946f684f3f741524aba3177f1aea737720bc19 Mon Sep 17 00:00:00 2001
From: Noam Lessner <nlessner@habana.ai>
Date: Mon, 20 Apr 2020 12:39:20 +0300
Subject: [PATCH] [SW-11862] goya2: force pcie recovery via gpio02

Add efuse that will allow ROM to force entering PCIe
recovery controlled via state of GPIO02

Change-Id: Icacbd81866498d9913ed524088c93a0bc9f9efd6
---

diff --git a/app/arcboot_rom.c b/app/arcboot_rom.c
index cc6b497..b3fcf34 100644
--- a/app/arcboot_rom.c
+++ b/app/arcboot_rom.c
@@ -104,10 +104,17 @@
 		arcboot_rom_set_efuse_read();
 }
 
+#define GPIO02_MASK BIT(2)
 static void arcboot_rom_in_recovery(void)
 {
+	uint8_t override = 1;
+
+	if ((arcboot_rom_global.config_fuse &
+	     EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL) == 0)
+		override = regmap_readl(mmPSOC_GPIO0_DATA) & GPIO02_MASK;
+
 	/* if recovery flag is set, bringup PCIe in recovery mode */
-	if (arcboot_rom_global.recovery_flag)
+	if (arcboot_rom_global.recovery_flag || (override == 0))
 		pcie_recovery();
 }
 
diff --git a/common/efuse.h b/common/efuse.h
index 9fc7b68..aa957c9 100644
--- a/common/efuse.h
+++ b/common/efuse.h
@@ -17,6 +17,7 @@
 #define EFUSE_SYSCFG_OFFSET			0x3A
 #define EFUSE_SECURITY_FW_VERSION_OFFSET	0x3B
 
+#define EFUSE_SYSCFG_GPIO02_RECOVERY_CONTROL	BIT(24)
 #define EFUSE_SYSCFG_WA_SNPS_CASE_01038891	BIT(23)
 #define EFUSE_SYSCFG_DISABLE_ROM_PCIE_RECOVERY	BIT(22)
 #define EFUSE_SYSCFG_DISABLE_SECONDARY_IMG	BIT(21)
